atef/widgets/happi.py

```diff
@@ -0,0 +1,470 @@
+        self._setup_ui()
+        # Set the client at the end, as this may trigger an update:
+        self.client = client
+
+    def _setup_ui(self):
+        """Configure UI elements at init time."""
+        self._setup_tree_view()
+        self._setup_list_view()
+
+        self.button_refresh.clicked.connect(self.refresh_happi)
+        self.list_or_tree_frame.layout().insertWidget(0, self.happi_list_view)
+
+        self.radio_by_name.clicked.connect(self._select_device_widget)
+        self.radio_by_category.clicked.connect(self._select_device_widget)
+        self.combo_by_category.currentTextChanged.connect(self._category_changed)
+        self.button_refresh.clicked.emit()
+
+    def _setup_list_view(self):
+        """Set up the happi_list_view."""
+        def list_selection_changed(
+            selected: QtCore.QItemSelection, deselected: QtCore.QItemSelection
+        ):
+            self.happi_items_selected.emit([idx.data() for idx in selected.indexes()])
+
+        view = self.happi_list_view
+        view.selectionModel().selectionChanged.connect(
+            list_selection_changed
+        )
+
+        view.setContextMenuPolicy(QtCore.Qt.CustomContextMenu)
+        view.customContextMenuRequested.connect(
+            self._list_view_context_menu
+        )
+
+        def set_filter(text):
+            self.happi_list_view.proxy_model.setFilterRegExp(text)
+
+        self.edit_filter.textEdited.connect(set_filter)
+
+    def _setup_tree_view(self):
+        """Set up the happi_tree_view."""
+        view = self.happi_tree_view
+        view.setVisible(False)
+        view.groups = [
+            self.combo_by_category.itemText(idx)
+            for idx in range(self.combo_by_category.count())
+        ]
+        self.list_or_tree_frame.layout().insertWidget(0, view)
+
+        def tree_selection_changed(
+            selected: QtCore.QItemSelection, deselected: QtCore.QItemSelection
+        ):
+            items = [
+                idx.data() for idx in selected.indexes()
+                if idx.parent().data() is not None  # skip top-level items
+            ]
+            self.happi_items_selected.emit(items)
+
+        view.selectionModel().selectionChanged.connect(
+            tree_selection_changed
+        )
+
+        view.setContextMenuPolicy(QtCore.Qt.CustomContextMenu)
+        view.customContextMenuRequested.connect(self._tree_view_context_menu)
+
+        def set_filter(text):
+            self.happi_tree_view.proxy_model.setFilterRegExp(text)
+
+        self.edit_filter.textEdited.connect(set_filter)
+        view.proxy_model.setRecursiveFilteringEnabled(True)
+
+    def _tree_view_context_menu(self, pos: QtCore.QPoint) -> None:
+        """Context menu for the happi tree view."""
+        self.menu = QtWidgets.QMenu(self)
+        index: QtCore.QModelIndex = self.happi_tree_view.indexAt(pos)
+        if index is not None:
+            def copy(*_):
+                copy_to_clipboard(index.data())
```

Comment:
The clipboard stuff here will be helpful in future contexts outside the scope of the config widget

Reply:
I took this out of my pydm PR and fixed it up further here. Agreed - it probably should go in pcdsutils.qt or something.